### PR TITLE
Go and Python codegen support symbols with hyphens in their names

### PR DESCRIPTION
**Pull request by mattolenik**
# Description

This change enables the fix for https://github.com/pulumi/crd2pulumi/issues/43, wherein hyphens `-` were mishandled and led to generation of invalid Go and Python. For example, code such as `func My-Thing() {}` or `class My-Thing`.

This change improves the existing casing functions and extracts them to a new subpackage, `cgstrings`.

Resolves https://github.com/pulumi/crd2pulumi/issues/43. When this change is merged `go get -u` will fix the crd2pulumi issue.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by iwahbe — APPROVED** [review]
This looks good to me.

File: pkg/codegen/cgstrings/cgstrings.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+// package cgstrings has various string processing functions that are useful during code generation.
+package cgstrings
+
+import (
+	"strings"
+	"unicode"
+)
+
+// Unhyphenate removes all hyphens from s, replacing them with r, then uppercasing the letter following each hyphen.
```
Comment (Frassle) [inline pkg/codegen/cgstrings/cgstrings.go:9]: "with r"?

**Comment by mikhailshilkov** [comment]
@mattolenik @iwahbe Should we modify some programgen test cases to cover changes from this PR?

File: pkg/codegen/cgstrings/cgstrings.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		if unicode.IsLower(r) {
+			res = append(res, runes[i:]...)
+			break
+		}
+		res = append(res, unicode.ToLower(r))
+	}
+	return string(res)
+}
+
+// UppercaseFirst uppercases the first letter of s.
```
Comment (Frassle) [inline pkg/codegen/cgstrings/cgstrings.go:33]: This is called Title elsewhere

File: pkg/codegen/go/utilities.go [hunk 3]
```diff
@@ -102,3 +103,7 @@ func makeSafeEnumName(name, typeName string) (string, error) {
 
 	return safeName, nil
 }
+
+func camel(s string) string {
```
Comment (Frassle) [inline pkg/codegen/go/utilities.go:107]: Can we just fix up the code to call cgstrings.Camel directly

File: pkg/codegen/python/utilities.go [hunk 4]
```diff
@@ -185,3 +186,17 @@ func pypiVersion(v semver.Version) string {
...
+	var underscores string
+	noUnderscores := strings.TrimLeftFunc(s, func(r rune) bool {
+		if r != '_' {
+			return false
+		}
+		underscores += "_"
+		return true
+	})
+	c := cgstrings.Camel(noUnderscores)
+	return underscores + strings.ToUpper(c[:1]) + c[1:]
```
Comment (Frassle) [inline pkg/codegen/python/utilities.go:201]:
```suggestion
	return underscores + cgstrings.Title(c)
```

**Comment by Frassle** [comment]
> Should we modify some programgen test cases to cover changes from this PR?

Or add new ones

File: pkg/codegen/go/utilities.go (hunk 3 above)
Comment (mattolenik) [reply pkg/codegen/go/utilities.go:107]: Sure, there were just two dozen or so instances of `camel()` so the change was just smaller this way

File: pkg/codegen/cgstrings/cgstrings.go (hunk 1 above)
Comment (mattolenik) [reply pkg/codegen/cgstrings/cgstrings.go:9]: Old comment fragment, will remove

File: pkg/codegen/cgstrings/cgstrings.go (hunk 2 above)
Comment (Frassle) [reply pkg/codegen/cgstrings/cgstrings.go:33]: That's fair enough reasoning to leave it as is

**Comment by mikhailshilkov** [comment]
@iwahbe @AaronFriel should we take this PR over and bring it to the finish line?

File: pkg/codegen/go/utilities.go (hunk 3 above)
Comment (iwahbe) [reply pkg/codegen/go/utilities.go]: Done in https://github.com/pulumi/pulumi/pull/10738/commits/b034765b08d9d333e3c7f299b6399fd0aab9c128

**Comment by AaronFriel** [comment]
Convert to draft & remove reviewers until ready for review?

**Comment by iwahbe** [comment]
> Convert to draft & remove reviewers until ready for review?

It is ready for review now. There was very little left to do.

**Comment by iwahbe** [comment]
> Convert to draft & remove reviewers until ready for review?

It should be ready for review now. @mattolenik addressed most of the feedback he got. I cleaned up the rest and regenerated the test case. It can be reviewed now.

File: pkg/codegen/python/utilities.go (hunk 4 above)
Comment (iwahbe) [reply pkg/codegen/python/utilities.go]: This turned into `cgstrings.UppercaseFirst(c)`.

**Review by Frassle — APPROVED** [review]

**Pull request merged**